# Loading all 13 NetCDF tile files at once

## Objectives:

Introduce a method for loading all 13 ECCO v4 NetCDF tile files at once.  


## Introduction

So far we've loaded a single NetCDF tile file at a time.  The `load_all_tiles_from_netcdf` method builds on the `load_tile_from_netcdf` routine, loading all 13 tiles at once.

Let's jump right in and load the 13 grid tile files.

### Loading an ECCO v4 netCDF grid tile file using  `load_all_tiles_from_netcdf`

Let's use `load_all_tiles_from_netcdf` to load all 13 grid tiles at once.  Call the new `Dataset` object `grid_all`

In [1]:
import matplotlib.pylab as plt
import numpy as np
import sys
import xarray as xr
from copy import deepcopy 
import ecco_v4_py as ecco

In [2]:
grid_dir='/Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/'
var = 'GRID'
var_type = 'grid'
grid_all = ecco.load_all_tiles_from_netcdf(grid_dir, var, var_type)
# minimize the metadata (optional)
ecco.minimal_metadata(grid_all)


>>> LOADING TILES FROM NETCDF

loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0001.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0002.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0003.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0004.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0005.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0006.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0007.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0008.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0009.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0010.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0011.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0012.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_grid/GRID.0013.nc
total file load time  0.768155097961 s
concatenated all tiles.  this can take a few minutes....
finished concatenating.  time = 0.53354382515 s
Remov

In [3]:
grid_all

<xarray.Dataset>
Dimensions:  (i: 90, i_g: 90, j: 90, j_g: 90, k: 50, k_g: 50, tile: 13)
Coordinates:
  * k        (k) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
  * i        (i) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
  * j        (j) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
  * i_g      (i_g) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 ...
  * j_g      (j_g) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 ...
  * k_g      (k_g) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 ...
  * tile     (tile) int64 1 2 3 4 5 6 7 8 9 10 11 12 13
Data variables:
    XC       (tile, j, i) float64 -111.6 -111.3 -110.9 -110.5 -110.0 -109.3 ...
    YC       (tile, j, i) float64 -88.24 -88.38 -88.52 -88.66 -88.8 -88.94 ...
    RAC      (tile, j, i) float64 3.623e+08 3.633e+08 3.643e+08 3.651e+08 ...
    Depth    (tile, j, i) float64 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ...
    AngleCS  (tile, j, i) float64 0.0615

### Examining the Dataset object contents


#### 1. Dimensions
`Dimensions:  (i: 90, i_g: 90, j: 90, j_g: 90, k: 50, k_g: 50, tile: 13)`

The *Dimensions* list now lists an extra dimension, **tile**.

#### 2. Coordinates
```
Coordinates:
  * k        (k) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
  * i        (i) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
  * j        (j) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
  * i_g      (i_g) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 ...
  * j_g      (j_g) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 ...
  * k_g      (k_g) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 ...
  * tile     (tile) int64 1 2 3 4 5 6 7 8 9 10 11 12 13
``` 

The **tile** coordinate is now an array, 1 .. 13


#### 3. Data Variables
```
Data variables:
    XC       (tile, j, i) float64 -111.6 -111.3 -110.9 -110.5 -110.0 -109.3 ...
    hFacC    (tile, k, j, i) float64 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ...
    XG       (tile, j_g, i_g) float64 -115.0 -115.0 -115.0 -115.0 -115.0 ...
    DXC      (tile, j, i_g) float64 1.558e+04 1.559e+04 1.559e+04 1.56e+04 ...
    hFacW    (tile, k, j, i_g) float64 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ...
    DYC      (tile, j_g, i) float64 1.156e+04 1.159e+04 1.162e+04 1.165e+04 ...
    hFacS    (tile, k, j_g, i) float64 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ...
    RC       (k) float64 -5.0 -15.0 -25.0 -35.0 -45.0 -55.0 -65.0 -75.0 ...
    RF       (k_g) float64 0.0 -10.0 -20.0 -30.0 -40.0 -50.0 -60.0 -70.0 ...
```

We now see that each *Data variable* that varies in the horizontal now has an extra dimension, **tile**. 

> **Note:** The ordering of the arrays is (tile, k, j, i)

## A closer look at the DataArray after loading with `load_all_tiles_from_netcdf`

Let's take a quick look at one of the variables loaded using the `load_all_tiles_from_netcdf` routine:

In [4]:
grid_all.XC

<xarray.DataArray 'XC' (tile: 13, j: 90, i: 90)>
array([[[-111.606468, -111.303001, ...,   64.805206,   64.819168],
        [-104.819603, -103.928444, ...,   64.410118,   64.4524  ],
        ..., 
        [ -37.5     ,  -36.5     , ...,   50.5     ,   51.5     ],
        [ -37.5     ,  -36.5     , ...,   50.5     ,   51.5     ]],

       [[ -37.5     ,  -36.5     , ...,   50.5     ,   51.5     ],
        [ -37.5     ,  -36.5     , ...,   50.5     ,   51.5     ],
        ..., 
        [ -37.5     ,  -36.5     , ...,   50.5     ,   51.5     ],
        [ -37.5     ,  -36.5     , ...,   50.5     ,   51.5     ]],

       ..., 
       [[-127.5     , -127.5     , ..., -127.5     , -127.5     ],
        [-126.5     , -126.5     , ..., -126.5     , -126.5     ],
        ..., 
        [ -39.5     ,  -39.5     , ...,  -39.5     ,  -39.5     ],
        [ -38.5     ,  -38.5     , ...,  -38.5     ,  -38.5     ]],

       [[-127.5     , -127.5     , ..., -115.505669, -115.166985],
        [-126.5    

The $XC$ `DataArray` object has **tile** has a new dimension of length 13.

## Loading non-grid ECCO variables using `load_all_tiles_from_netcdf`

As with `load_tile_from_netcdf` we can give our variables better labeled coordinates by specifying their c-grid point category.

We'll demonstrate by loading three variables, one each that are on 'c','u', and 'v' points 

### A 'c' point variable: SSH

Let's load tile 3 of the 'c' point sea surface height (SSH) variable.  

In [5]:
data_dir='/Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/'    
var = 'SSH'
var_type = 'c'
ssh_all = ecco.load_all_tiles_from_netcdf(data_dir, var, var_type)
ecco.minimal_metadata(ssh_all)


>>> LOADING TILES FROM NETCDF

loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0001.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0002.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0003.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0004.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0005.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0006.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0007.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0008.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0009.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0010.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0011.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0012.nc
loading /Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/SSH/SSH.0013.nc
total file load time  0.506529092789 s
concatenated all tiles.  this ca

Since this is the first time we're loading an output variable from the state estimate let's closely examine `ssh_tile_3`.

In [6]:
ssh_all

<xarray.Dataset>
Dimensions:   (i: 90, j: 90, tile: 13, time: 288)
Coordinates:
  * time      (time) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 ...
  * j         (j) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
  * i         (i) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
    tim       (time) datetime64[ns] 1992-01-16 1992-02-16 1992-03-16 ...
    timestep  (time) float64 732.0 1.428e+03 2.172e+03 2.892e+03 3.636e+03 ...
    lon_c     (tile, j, i) float64 -111.6 -111.3 -110.9 -110.5 -110.0 -109.3 ...
    lat_c     (tile, j, i) float64 -88.24 -88.38 -88.52 -88.66 -88.8 -88.94 ...
  * tile      (tile) int64 1 2 3 4 5 6 7 8 9 10 11 12 13
Data variables:
    SSH       (time, tile, j, i) float64 nan nan nan nan nan nan nan nan nan ...
Attributes:
    description:    SSH -- ECCO v4 ocean state estimate, release 3 -- 1992-2015
    date:           02-May-2017
    Conventions:    CF-1.6
    _FillValue:     nan
    missing_value:  nan

Notice that unlike the grid files that we've loaded up until now, the `ssh_tile_3` `Dataset` object only has one *Data variable*, $SSH$.  Non-grid ECCO v4 NetCDF tile files only ever store *one* physical variable. Consequently, the *dimensions* of the `Dataset` will be the same as the *dimensions* of the single *data variable*.    Let's take a look at the $SSH$ `DataArary`:

In [7]:
ssh_all.SSH

<xarray.DataArray 'SSH' (time: 288, tile: 13, j: 90, i: 90)>
array([[[[      nan, ...,       nan],
         ..., 
         [-1.483355, ..., -1.354059]],

        ..., 
        [[-0.713382, ...,       nan],
         ..., 
         [-1.46542 , ...,       nan]]],


       ..., 
       [[[      nan, ...,       nan],
         ..., 
         [-1.408568, ..., -1.319521]],

        ..., 
        [[-0.580478, ...,       nan],
         ..., 
         [-1.390469, ...,       nan]]]])
Coordinates:
  * time      (time) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 ...
  * j         (j) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
  * i         (i) float64 1.0 2.0 3.0 4.0 5.0 6.0 7.0 8.0 9.0 10.0 11.0 12.0 ...
    tim       (time) datetime64[ns] 1992-01-16 1992-02-16 1992-03-16 ...
    timestep  (time) float64 732.0 1.428e+03 2.172e+03 2.892e+03 3.636e+03 ...
    lon_c     (tile, j, i) float64 -111.6 -111.3 -110.9 -110.5 -110.0 -109.3 ...
    lat_c     (tile, j, i) float64 -

#### Dimensional coordinates
As expected, the `SSH` `DataArray` uses **i, j** coordinates for its horizontal dimensions.  We also see a **tile** and **time** coordinates.  The ordering of the $SSH$ dimensions is **time, tile, j, i** with the logic being that **tile** is a kind of space coordinate.

We find 288 records in the **time** dimensions.  One for each month of the 1992-2015 state estimate.  There is one record in the **tile** dimensions because we have only loaded a single tile file so far.

#### Non-dimensional coordinates

Notice the four *Coordinates* that do not have an "\*" in front of their names.  These are so-called [non-dimensional coordinates](http://xarray.pydata.org/en/stable/data-structures.html#coordinates).  Think of non-dimensional coordinates as helpful extra coordinate labels.  Here, the non-dimensional coordinates include two for the **time** dimension: **tim** and **timestep**, and two for the space dimensions: **lon_c, lat_c**.  

##### The time non-dimensional coordinates

In [8]:
ssh_tile_3.time

NameError: name 'ssh_tile_3' is not defined

The **tim** non-dimensional coordinate provides the calendar date (centered in the month) for the **time** monthly index and the **timestep** non-dimensional coordinate provides the the timestep number corresponding to each month (the ECCO v4 simulation uses 1-hourly timesteps).

##### The space non-dimensional coordinates
**lon_c** and **lat_c** non-dimensional coordinates provide the longitude and latitude for each of the 'c' points of $SSH$.

In [ ]:
ssh_tile_3.lon_c

### A 'u' point variable: UVEL

Let's load tile 3 of the 'u' point horizontal velocity in the local $x$ direction variable, $UVEL$.

In [ ]:
data_dir='/Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/UVEL/'    
var = 'UVEL'
var_type = 'u'
tile_index = 3
uvel_tile_3 = ecco.load_tile_from_netcdf(data_dir, var, var_type, tile_index)

Let's look at `uvel_tile_3`.  This time let's also remove some of the descriptive NetCDF file *Attributes* using a little routine called `minimal_metadata`.  We've already seen these attributes a number of times.

In [ ]:
ecco.minimal_metadata(uvel_tile_3)

In [ ]:
uvel_tile_3

`uvel_tile_3` has one *Data variable*, $UVEL$.  Let's take a look at the $UVEL$ `DataArary`:

In [ ]:
uvel_tile_3.UVEL

#### Dimensional coordinates
As expected, $UVEL$ uses the **i_g, j** coordinates for its horizontal dimensions. Unlike $SSH$, $UVEL$ has three-dimensions in space so we find a **k** coordinate.  The ordering of the three-dimensional ECCO v4 output is **time, tile, k, j, i**.

#### Non-dimensional coordinates

$UVEL$ has one new non-dimensional coordinate for **k**: **dep** and two new non-dimensional coordinates for space: **lon_u, lat_u**

##### The **dep** non-dimensional coordinates

**dep** is the depth of the center of the tracer grid cell in meters:

In [ ]:
uvel_tile_3.dep

##### The space non-dimensional coordinates
**lon_u** and **lat_u** provide the longitude and latitude for each **i_g,j** point.

### A 'v' point variable: VVEL

Let's load tile 3 of the 'v' point horizontal velocity in the local $x$ direction variable, $VVEL$.

In [ ]:
data_dir='/Volumes/ECCO_BASE/ECCO_v4r3/nctiles_monthly/VVEL/'    
var = 'VVEL'
var_type = 'v'
tile_index = 3
vvel_tile_3 = ecco.load_tile_from_netcdf(data_dir, var, var_type, tile_index)
ecco.minimal_metadata(vvel_tile_3)

Let's look at the *Coordinates* of $VVEL$.

In [ ]:
vvel_tile_3.coords

#### Dimensional coordinates
As expected, `vvel_tile_3` uses the **i, j_g** coordinates for its horizontal dimensions. 

#### Non-dimensional coordinates

$VVEL$ has two new non-dimensional coordinates for space: **lon_v, lat_v**.  As you might expect, these are the longitude and latitude of this v-point variable.

## Conclusion

Now you know how to load variables so that their coordinates correspond with their location on the c-grid.